### PR TITLE
[Fallback] Initial version of the interop test.

### PR DESCRIPTION
**Pull request by eugeneo**
I suggest merging this as soon as possible and then make smaller pull requests to better integrate new test with the framework.

There will be another PR that adds Go control panel docker image.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -210,7 +210,7 @@ pip install -r requirements.lock
 # Generate protos
 python -m grpc_tools.protoc --proto_path=. \
   --python_out=. --grpc_python_out=. --pyi_out=. \
-  protos/grpc/testing/*.proto
+  protos/grpc/testing/*.proto protos/grpc/testing/**/*.proto
```
Comment (sergiitk) [inline README.md:213]: we need to add this to https://github.com/grpc/psm-interop/blob/657acc2637b34ea167c5289762f02be07dae7042/.github/workflows/psm-interop.yaml#L49-L55

File: framework/helpers/docker.py [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+from mako.template import Template
+
+from protos.grpc.testing import messages_pb2
+from protos.grpc.testing import test_pb2_grpc
+from protos.grpc.testing.xdsconfig import control_pb2
+from protos.grpc.testing.xdsconfig import service_pb2_grpc
+
+
+class Bootstrap:
+    def __init__(self, base: str, ports: List[int], host_name: str):
```
Comment (sergiitk) [inline framework/helpers/docker.py:24]: nit: we can use `list[]` typing notation now; no extra imports needed
```suggestion
    def __init__(self, base: str, ports: list[int], host_name: str):
```

File: framework/helpers/docker.py [hunk 3]
```diff
@@ -0,0 +1,308 @@
+from datetime import datetime
+from datetime import timedelta
+from math import ceil
+from pathlib import Path
+from pathlib import PosixPath
+from queue import Queue
+from threading import Thread
```
Comment (sergiitk) [inline framework/helpers/docker.py:7]: here and elsewhere:

> Use import statements for packages and modules only, not for individual types, classes, or functions.
> — https://google.github.io/styleguide/pyguide.html#22-imports

File: framework/helpers/docker.py [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+        self.__queue.put(ChildProcessEvent(source, message))
+
+    def mount_dir(self):
+        return self.__bootstrap.mount_dir()
+
+
+def _Sanitize(l: str) -> str:
+    if l.find("\0") < 0:
+        return l
+    return l.replace("\0", "�")
```
Comment (sergiitk) [inline framework/helpers/docker.py:119]: What's `�`?

File: README.md (hunk 1 above)
Comment (sergiitk) [inline README.md:213]:
```suggestion
  $(find protos/grpc/testing/ -type f -name '*.proto')
```
I don't like bash globbing - it only gets results people expect when `shopt -s globstar` is enabled. Let's replace it with `find` instead.

File: framework/helpers/docker.py (hunk 3 above)
Comment (eugeneo) [reply framework/helpers/docker.py]: Done

File: framework/helpers/docker.py (hunk 4 above)
Comment (eugeneo) [reply framework/helpers/docker.py:115]: Just a placeholder. Need to find why our code logs \0. I opened a Core bug and assigned to myself.

File: README.md (hunk 1 above)
Comment (eugeneo) [reply README.md]: Replaced with the dir name. This is readme, need to make it easy to understand and follow

**Review by eugeneo — COMMENTED** [review]
Thank you for the review. PTAL.

File: framework/helpers/docker.py (hunk 3 above)
Comment (eugeneo) [reply framework/helpers/docker.py]: I do not know how to comply with linters other then importing individual protobuf generated classes. I made a change.

File: protos/grpc/testing/xdsconfig/control.proto [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package grpc.testing.xdsconfig;
```
Comment (sergiitk) [inline protos/grpc/testing/xdsconfig/control.proto:17]: Quick question: will these protos eventually be added to the `xdsconfig` folder of https://github.com/grpc/grpc-proto/tree/master/grpc/testing?

Comment (eugeneo) [reply protos/grpc/testing/xdsconfig/control.proto:17]: They will be significantly reworked and added there. There are plans to use control plane for other test cases. This is just an initial commit.

File: tests/fallback_test.py [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+
+
+def GetFreePort() -> int:
+    with (sock := socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as sock:
+        sock.bind(("localhost", 0))
+        return sock.getsockname()[1]
+
+
+class FallbackTest(unittest.TestCase):
+    __bootstrap: Bootstrap = None
```
[the remark this answers is not in the document]
Comment (sergiitk) [reply tests/fallback_test.py]: I wonder if pylint caches unused variables like this. If no, we should set it up (not in this PR, of course).

File: framework/helpers/docker.py (hunk 3 above)
Comment (sergiitk) [reply framework/helpers/docker.py]: Here's how:
```diff
- from protos.grpc.testing.xdsconfig.control_pb2 import StopOnRequestRequest
- from protos.grpc.testing.xdsconfig.control_pb2 import UpsertResourcesRequest
+ from protos.grpc.testing.xdsconfig import control_pb2
```

Then just use `control_pb2.StopOnRequestRequest` and `control_pb2.UpsertResourcesRequest`.

File: framework/helpers/docker.py (hunk 4 above)
Comment (sergiitk) [reply framework/helpers/docker.py:116]: Can you link the bug here please? Was it resolved?

File: README.md (hunk 1 above)
Comment (sergiitk) [reply README.md]: Ok, looks good.

File: protos/grpc/testing/xdsconfig/control.proto (hunk 5 above)
Comment (sergiitk) [reply protos/grpc/testing/xdsconfig/control.proto:17]: Sounds good, thanks for explaining. Is there a tracking ticket for moving these protos to grpc/grpc-proto?

File: framework/helpers/docker.py [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+        template = mako.template.Template(filename="templates/bootstrap.json")
+        file = template.render(
+            servers=[f"{self.host_name}:{port}" for port in self.ports]
+        )
+        destination = self.get_mount_dir() / "bootstrap.json"
+        with open(destination, "w", encoding="utf-8") as f:
+            f.write(file)
+            absl.logging.debug("Generated bootstrap file at %s", destination)
+
+    def MakeWorkingDir(self, base: str):
```
Comment (sergiitk) [inline framework/helpers/docker.py:38]: Here and elsewhere: use `snake_case`. The only exception from it is classes derived from unittest.Testcase that require to override methods like `setUpClass`, etc.

File: framework/helpers/docker.py [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+
+    def __repr__(self) -> str:
+        return f"data={self.data}, source={self.source}"
+
+
+class ProcessManager:
+    def __init__(
+        self,
+        bootstrap: Bootstrap,
+        nodeId: str,
```
Comment (sergiitk) [inline framework/helpers/docker.py:71]: same as with method names, use `snake_case` for parameters and variables.

File: framework/helpers/docker.py [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+        verbosity="info",
+    ):
+        self.dockerClient = docker.DockerClient.from_env()
```
Comment (sergiitk) [inline framework/helpers/docker.py:74]: Tip on fixing pylint no-member issues: figure out in what file the class is defined, and import it. In this case, you want

```
import docker.client
```

And then refer refer to `DockerClient` as `docker.client.DockerClient`.

**Review by sergiitk — CHANGES_REQUESTED** [review]
More notes and suggestions how to pass the linter.

File: tests/fallback_test.py [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+    @staticmethod
+    def setUpClass():
+        FallbackTest.bootstrap = framework.helpers.docker.Bootstrap(
+            framework.helpers.logs.log_dir_mkdir("bootstrap"),
+            ports=[GetFreePort() for _ in range(2)],
+            host_name=FLAGS.host_name,
+        )
+
+    def setUp(self):
+        absl.logging.info(f"Starting %s", self.id())
```
[the remark this answers is not in the document]
Comment (eugeneo) [reply tests/fallback_test.py]: Done

**Review by eugeneo — COMMENTED** [review]
Thanks for the review! I addressed the comments.

File: tests/fallback_test.py [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+            node_id=_NODE_ID.value,
+        )
+
+    def start_client(self, port: int = None, name: str = None):
+        logger.debug("Starting client process")
+        if name is None:
+            name = framework.xds_flags.CLIENT_NAME.value
+        return framework.helpers.docker.Client(
+            manager=self.process_manager,
+            name=name,
```
Comment (XuanWang-Amos) [inline tests/fallback_test.py:61]: Can be just `name=name or framework.xds_flags.CLIENT_NAME.value`

File: tests/fallback_test.py [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+            upstream=f"{_HOST_NAME.value}:{upstream_port}",
+            image=_CONTROL_PLANE_IMAGE.value,
+        )
+
+    def start_server(self, name: str, port: int = None):
+        logger.debug('Starting server "%s"', name)
+        port = get_free_port() if port is None else port
+        return framework.helpers.docker.GrpcProcess(
+            name=name,
+            port=port,
```
Comment (XuanWang-Amos) [inline tests/fallback_test.py:83]: Dito, `port=port or get_free_port()`

**Review by eugeneo — COMMENTED** [review]
Thanks for the reviews! I addressed the comments

File: framework/helpers/docker.py (hunk 3 above)
Comment (eugeneo) [reply framework/helpers/docker.py]: Done

File: tests/fallback_test.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (eugeneo) [reply tests/fallback_test.py]: Ack

File: framework/helpers/docker.py [hunk 13]
```diff
@@ -0,0 +1,331 @@
...
+import datetime
+import logging
+import math
+import pathlib
+import queue
+import threading
+
+from docker.client import DockerClient
+from docker.errors import NotFound
+from docker.types import ContainerConfig
```
Comment (sergiitk) [inline framework/helpers/docker.py:11]: As discussed before, we don't import classes from files: https://google.github.io/styleguide/pyguide.html#224-decision

The only exceptions from this are:
- typing module
- collections.abc module
- typing_extensions module

Comment (eugeneo) [reply framework/helpers/docker.py]: Frankly speaking, I do not quite understand all those different import types in Python. I addressed this comment as I understood it.

File: tests/fallback_test.py [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+        ):
+            self.assertTrue(
+                client.expect_message_in_output(
+                    "UNAVAILABLE: xDS channel for server"
+                )
+            )
+            self.assertEqual(client.get_stats(5).num_failures, 5)
+            # Fallback control plane start, send traffic to server2
+            with self.start_control_plane(
+                "fallback_xds_config", 1, server2.port
```
Comment (XuanWang-Amos) [inline tests/fallback_test.py:102]: Can we add key to the params like we did for `primary_xds_config`?

File: tests/fallback_test.py [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+                ):
+                    self.assertTrue(
+                        client.expect_message_in_output(
+                            "parsed Cluster example_proxy_cluster", timeout_s=30
+                        )
+                    )
+                    stats = client.get_stats(10)
+                    self.assertEqual(stats.num_failures, 0)
+                    self.assertIn("server1", stats.rpcs_by_peer)
+                    self.assertGreater(stats.rpcs_by_peer["server1"], 0)
```
Comment (XuanWang-Amos) [inline tests/fallback_test.py:121]: Will all RPCs going to `server1` once it's up? If they are, we probably want to check no RPC is going to `server2` (same questions for all test cases).

File: tests/fallback_test.py [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+                # Primary control plane down
+                stats = client.get_stats(5)
+                self.assertEqual(stats.num_failures, 0)
+                self.assertEqual(stats.rpcs_by_peer["server1"], 5)
```
Comment (XuanWang-Amos) [inline tests/fallback_test.py:125]: Why it's `server1` here?

Comment (eugeneo) [reply tests/fallback_test.py:127]: Cached value used if we have a consistent config that we got from the control plane before it went down. I updated the comment.

File: tests/fallback_test.py (hunk 15 above)
Comment (eugeneo) [reply tests/fallback_test.py:123]: Some will still arrive to server2 while the primary config is being downloaded. Other checks verify that all RPCs go to `server1` so no need to check that none arrived to `server2`

File: tests/fallback_test.py (hunk 14 above)
Comment (eugeneo) [reply tests/fallback_test.py]: Done.

**Comment by sergiitk** [comment]
> @sergiitk sergiitk self-requested a review now
I think this dismissed my last review status.

@XuanWang-Amos - feel free to finish the review and approve when you think it's ready.

**Review by XuanWang-Amos — APPROVED** [review]
LGTM, sorry for the delay!

**Pull request merged**